server/src/main/java/org/opensearch/index/remote/RemoteMigrationClusterStateUtils.java

```diff
@@ -0,0 +1,191 @@
+ */
+
+package org.opensearch.index.remote;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.Version;
+import org.opensearch.cluster.ClusterState;
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.cluster.metadata.MetadataCreateIndexService;
+import org.opensearch.cluster.node.DiscoveryNodes;
+import org.opensearch.cluster.routing.IndexRoutingTable;
+import org.opensearch.cluster.routing.RoutingTable;
+import org.opensearch.cluster.routing.ShardRouting;
+import org.opensearch.cluster.routing.ShardRoutingState;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.index.remote.RemoteStoreEnums.PathHashAlgorithm;
+import org.opensearch.index.remote.RemoteStoreEnums.PathType;
+import org.opensearch.indices.replication.common.ReplicationType;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.opensearch.cluster.metadata.IndexMetadata.REMOTE_STORE_CUSTOM_KEY;
+import static org.opensearch.indices.RemoteStoreSettings.CLUSTER_REMOTE_STORE_PATH_HASH_ALGORITHM_SETTING;
+import static org.opensearch.indices.RemoteStoreSettings.CLUSTER_REMOTE_STORE_PATH_TYPE_SETTING;
+
+/**
+ * Utils for checking and mutating cluster state during remote migration
+ *
+ * @opensearch.internal
+ */
+public class RemoteMigrationClusterStateUtils {
+    private static final Logger logger = LogManager.getLogger(RemoteMigrationClusterStateUtils.class);
+
+    /**
+     * During docrep to remote store migration, applies the following remote store based index settings
+     * once all shards of an index have moved over to remote store enabled nodes
+     * <br>
+     * Also appends the requisite Remote Store Path based custom metadata to the existing index metadata
+     */
+    public static void maybeAddRemoteIndexSettings(
+        IndexMetadata indexMetadata,
+        IndexMetadata.Builder indexMetadataBuilder,
+        RoutingTable routingTable,
+        String index,
+        DiscoveryNodes discoveryNodes,
+        String segmentRepoName,
+        String tlogRepoName
+    ) {
+        Settings currentIndexSettings = indexMetadata.getSettings();
+        if (needsRemoteIndexSettingsUpdate(routingTable.indicesRouting().get(index), discoveryNodes, currentIndexSettings)) {
+            logger.info(
+                "Index {} does not have remote store based index settings but all primary shards and STARTED replica shards have moved to remote enabled nodes. Applying remote store settings to the index",
+                index
+            );
+            Settings.Builder indexSettingsBuilder = Settings.builder().put(currentIndexSettings);
+            MetadataCreateIndexService.updateRemoteStoreSettings(indexSettingsBuilder, segmentRepoName, tlogRepoName);
+            indexMetadataBuilder.settings(indexSettingsBuilder);
+            indexMetadataBuilder.settingsVersion(1 + indexMetadata.getVersion());
```

Comment:
do we need it on path type update as well ? What about both updates at same time - would it bump it by 2 ?

Reply:
We do not update the `Settings` during the path metadata update. This counter is only meant for bumping up the version for `Settings` if there is an update on that. There is an assertion here which would trip otherwise

https://github.com/opensearch-project/OpenSearch/blob/c9f1565939c8328c23f235aa547ab55d0085af73/server/src/main/java/org/opensearch/index/IndexService.java#L918-L932